csharp/src/ParallelSelectExt.cs

```diff
@@ -0,0 +1,368 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+using JetBrains.Annotations;
+
+namespace ArmoniK.Utils;
+
+/// <summary>
+///   Options for ParallelSelect and ParallelWait
+/// </summary>
+[PublicAPI]
+public struct ParallelTaskOptions
+{
+  private readonly int parallelismLimit_ = 0;
+
+  /// <summary>Limit the parallelism for ParallelSelect and ParallelWait</summary>
+  public int ParallelismLimit
+  {
+    get
+      => parallelismLimit_ switch
+         {
+           < 0 => int.MaxValue,
+           0   => Environment.ProcessorCount,
+           _   => parallelismLimit_,
+         };
+    init => parallelismLimit_ = value;
+  }
+
+  /// <summary>Cancellation token used for stopping the enumeration</summary>
+  public CancellationToken CancellationToken { get; init; } = default;
+
+  /// <summary>
+  ///   Options for ParallelSelect and ParallelWait.
+  ///   If parallelismLimit is 0, the number of threads is used as the limit.
+  ///   If parallelismLimit is negative, no limit is enforced.
+  /// </summary>
+  /// <param name="parallelismLimit">Limit the parallelism</param>
+  /// <param name="cancellationToken">Cancellation token used for stopping the enumeration</param>
+  public ParallelTaskOptions(int               parallelismLimit,
+                             CancellationToken cancellationToken = default)
+  {
+    ParallelismLimit  = parallelismLimit;
+    CancellationToken = cancellationToken;
+  }
+
+  /// <summary>
+  ///   Options for ParallelSelect and ParallelWait.
+  ///   Parallelism is limited to the number of threads.
+  /// </summary>
+  public ParallelTaskOptions()
+    : this(0)
+  {
+  }
+
+  /// <summary>
+  ///   Options for ParallelSelect and ParallelWait.
+  ///   Parallelism is limited to the number of threads.
+  /// </summary>
+  /// <param name="cancellationToken">Cancellation token used for stopping the enumeration</param>
+  public ParallelTaskOptions(CancellationToken cancellationToken)
+    : this(0,
+           cancellationToken)
+  {
+  }
+}
+
+[PublicAPI]
+public static class ParallelSelectExt
+{
+  /// <summary>
+  ///   Iterates over the input enumerable and spawn multiple parallel tasks.
+  ///   The maximum number of tasks in flight at any given moment is given in the `parallelTaskOptions`.
+  ///   All results are collected in-order.
+  /// </summary>
+  /// <param name="enumerable">Enumerable to iterate on</param>
+  /// <param name="parallelTaskOptions">Options (eg: parallelismLimit, cancellationToken)</param>
+  /// <typeparam name="T">Return type of the tasks</typeparam>
+  /// <returns>Asynchronous results of tasks</returns>
+  [PublicAPI]
+  public static async IAsyncEnumerable<T> ParallelWait<T>(this IEnumerable<Task<T>> enumerable,
```

Comment:
I don't like the name.
If it is a wait semantic, it should be some kind of WhenXxxx. ParallelEvaluation?